R/util_dfc.R

```diff
@@ -0,0 +1,238 @@
+
+  # Sanity checks
+  if(!is.data.frame(df1) | !is.data.frame(df2)) {
+    stop("Error - at least one of the first two arguments for util_dfc.compare_colnames is not a data frame.")
+  }
+
+  return(util_dfc.compare_vecs(names(df1), names(df2)))
+}
+
+
+
+
+util_dfc.compare_identifiers <- function(df1, df2, id_cols, id_cols_uniquely_identify_rows = FALSE) {
+  # Compare two dfs on a set of one or more commonly-shared columns that identify rows.
+  # If rows are identified by multiple columns (e.g. team x student x week), include all column names in id_cols.
+  # Return a list with shared IDs, IDs in df1 only, and IDs in df2 only.
+  #
+  # If you expect id_cols to UNIQUELY identify each row in both dfs, set id_cols_uniquely_identify_rows to TRUE.
+  # Otherwise, duplicate ID values within each data frame will get ignored.
+  # Example use case: you want to compare a student-x-week df with a student df to see if they have the same students.
+  # The commonly-shared id_col is student ID.
+  # Since student ID doesn't uniquely identify rows in the first df, id_cols_uniquely_identify_rows = FALSE.
+
+  # Sanity checks
+  if(!is.data.frame(df1) | !is.data.frame(df2)) {
+    stop("Error - at least one of the first two arguments for util_dfc.compare_unique_identifiers is not a data frame.")
+  }
+  if(is.na(id_cols) | length(id_cols) == 0) {
+    stop("Error - id_cols argument is NA or empty.")
+  }
+  if(any(!id_cols %in% names(df1)) | any(!id_cols %in% names(df2))) {
+    stop("Error - not all id columns provided are in both data frames.")
+  }
+
+  # Extract the IDs from df1 and df2 in character format.
+  # If IDs are defined by multiple columns, use them all with "__" as a separator.
+  # Note: this section treats NAs as the character string "NA" in an ID.
+  df1_ids <- util_dfc.get_condensed_ids(df1, id_cols)
+  df2_ids <- util_dfc.get_condensed_ids(df2, id_cols)
+
+  # If the IDs aren't expected to uniquely identify rows, then cut duplicates from them.
+  if(!id_cols_uniquely_identify_rows) {
+    df1_ids <- df1_ids[!duplicated(df1_ids)]
+    df2_ids <- df2_ids[!duplicated(df2_ids)]
+  }
+
+  # Warn if the IDs still have duplicates in either df!
+  # This would mean that the user expected no duplicates but actually had them.
+  if(any(duplicated(df1_ids))) {
+    warning("Warning - first data frame passed to util_dfc.compare_unique_identifiers has unexpected duplicate ID values:")
+    util.print_pre(unique(df1_ids[duplicated(df1_ids)]))
+  }
+  if(any(duplicated(df2_ids))) {
+    warning("Warning - second data frame passed to util_dfc.compare_unique_identifiers has unexpected duplicate ID values:")
+    util.print_pre(unique(df2_ids[duplicated(df2_ids)]))
+  }
+
+  # use util_dfc.compare_vecs on id_cols of each df
+  return(util_dfc.compare_vecs(df1_ids, df2_ids))
+}
+
+
+
+
+util_dfc.compare_dfs <- function(df1, df2, id_cols = c()) {
+  # Big wrapper function for giving an overall picture of how two DFs differ:
```

Comment:
A high-level thought on this wrapper function: I'm not sure the gymnast package needs to have a wrapper function like this. I expect many projects to draw on these df comparison functions, not just the EP. And I think the set and order of comparison operations is likely to differ from project to project, or even from comparison to comparison within projects.

For example, I am 100% excited to use this function suite in the IES data pipeline! I think we will frequently want to ensure that although new identifiers (rows in combination) have been added, none have been subtracted. This suite offers a very elegant set of functions to do that, but I would likely use your compare_identifiers function a la carte, rather than as part of a wrapper function, for this purpose.

If we do decide we need a wrapper function, then I might suggest having it return an object including mismatched identifiers, missing column names, etc., rather than printing a report to the screen. When datasets don't match, you almost always need access to the offending identifiers, column names, etc. for the purposes of figuring out wtf went wrong (which is why I _love_ that your other functions do return objects with this information—and another part of why I'm not convinced we need a whole wrapper).

Reply:
Very helpful feedback. FWIW, I have now updated this wrapper function to return two things: a summary table of which COLUMNS have mismatched values and how many, and a summary table of which ROW IDENTIFIERS have mismatched values and how many. These value-by-value comparisons are only possible if you do extensive sorting and pre-cleaning, so a wrapper function might be necessary.